python/pdo/contract/replication.py

```diff
@@ -0,0 +1,277 @@
+# Copyright 2018 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import concurrent.futures
+import queue
+import threading
+
+from pdo.contract.state import ContractState
+import pdo.service_client.service_data.eservice as service_db
+
+import logging
+logger = logging.getLogger(__name__)
+
+# -----------------------------------------------------------------
+__replication_manager_executor__ = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+__pending_replication_tasks_manager_queue__ = queue.Queue()
+
+__replication_workers_executor__ = dict() #key is service id, value is a ThreadPoolExecutor object that manages the worker threads for this storage service
+num_threads_per_storage_service = 2 # we many want to parametrize this later
+__pending_replication_tasks_workers_queues__ = dict() #key is service id, value is a queue of pending replication tasks for this storage service
+
+__condition_variable_for_completed_tasks__ = threading.Condition() # used to notify the parent thread about a new task that got completed (if the parent is waiting)
```

Comment:
I think this `__condition_variable_for_completed_tasks__ ` works, though it should be in the replication request object.

Long answer.
As I understand it, there may be multiple tasks being processed by the manager and then by the workers.
However, once a replication request completes or fails there is one global condition variable which is updated.

Now, if you have 2 request from 2 threads, which are waiting for them to complete, when the first terminates you notify both of them -- though one will make progress, the other keeps waiting.